### PR TITLE
Add `LifecyclePolicy` field to `Repository`

### PR DESCRIPTION
**Pull request by A-Hilaly**
https://github.com/aws-controllers-k8s/community/issues/1043

Description of changes:
- Add `LifecyclePolicy` field to `Repository` resource
- e2e tests for updates or creation of repositories with lifecycle policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/e2e/resources/repository_lifecycle_policy.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+apiVersion: ecr.services.k8s.aws/v1alpha1
+kind: Repository
+metadata:
+  name: $REPOSITORY_NAME
+spec:
+  name: $REPOSITORY_NAME
+  imageScanningConfiguration:
+    scanOnPush: false
+  imageTagMutability: MUTABLE
+  lifecyclePolicy: '{"rules":[{"rulePriority":1,"description":"Expire images older than 14 days","selection":{"tagStatus":"untagged","countType":"sinceImagePushed","countUnit":"days","countNumber":14},"action":{"type":"expire"}}]}'
```
Comment (A-Hilaly) [inline test/e2e/resources/repository_lifecycle_policy.yaml:10]: Mmmm thinking that maybe i should make this a `$LIFECYCLE_POLICY`

File: apis/v1alpha1/ack-generate-metadata.yaml [hunk 2]
```diff
@@ -1,13 +1,13 @@
 ack_generate_info:
-  build_date: "2021-10-08T18:50:03Z"
-  build_hash: 385779a205bea50e8762b76bc75cab957cf723b9
-  go_version: go1.15
-  version: v0.15.1
-api_directory_checksum: 3c829828db77ef587929da78f41d52b2459fb054
+  build_date: "2021-11-12T16:27:36Z"
+  build_hash: 6ce1a672eabd3908bdaa4ace356e1b58ee3e80ba
+  go_version: go1.16.4
+  version: v0.15.2
```
Comment (A-Hilaly) [inline apis/v1alpha1/ack-generate-metadata.yaml:5]: Also, probably we'll need to ship #21 first

File: pkg/resource/repository/hook.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	"context"
+
+	svcapitypes "github.com/aws-controllers-k8s/ecr-controller/apis/v1alpha1"
+	ackrtlog "github.com/aws-controllers-k8s/runtime/pkg/runtime/log"
+	svcsdk "github.com/aws/aws-sdk-go/service/ecr"
+)
+
+// addPutFieldsToSpec will describe the fields that are not return by
+// DescribeRepository calls
+func (rm *resourceManager) describeRepositoryFields(
```
Comment (jaypipes) [inline pkg/resource/repository/hook.go:13]: How about we call this `setResourceAdditionalFields` since it is setting fields on the resource.

**Review by jaypipes — CHANGES_REQUESTED** [review]
One tiny little naming nit otherwise this looks fantastic. Nice work @A-Hilaly!

File: apis/v1alpha1/ack-generate-metadata.yaml (hunk 2 above)
Comment (A-Hilaly) [reply apis/v1alpha1/ack-generate-metadata.yaml:5]: shipped

File: pkg/resource/repository/hook.go (hunk 3 above)
Comment (A-Hilaly) [reply pkg/resource/repository/hook.go]: Agreed. Changed to `setResourceAdditionalFields`

**Review by jaypipes — APPROVED** [review]
:+1:  thanks for renaming that one method.

**Pull request merged**